h/static/scripts/annotator/guest.coffee

```diff
@@ -211,21 +161,131 @@ module.exports = class Guest extends Annotator
+        normedRange = range.normalize(root)
+
+        highlights = highlighter.highlightRange(normedRange)
+        rect = highlighter.getBoundingClientRect(highlights)
+
+        $(highlights).data('annotation', anchor.annotation)
+
+        anchor.highlights = highlights
+        anchor.pos =
+          left: rect.left + window.scrollX
+          top: rect.top + window.scrollY
+
+        return anchor
+
+    sync = (anchors) ->
+      annotation.$anchors = ({pos} for {pos} in anchors)
+      annotation.$orphan = anchors.length > 0
+      for anchor in anchors
+        if anchor.range?
+          annotation.$orphan = false
+
+      self.anchors = self.anchors.concat(anchors)
+      self.plugins.BucketBar?.update()
+      self.plugins.CrossFrame?.sync([annotation])
+
+    for anchor in self.anchors.splice(0, self.anchors.length)
+      if anchor.annotation is annotation
+        if anchor.range? and anchor.target in annotation.target
+          anchors.push(anchor)
+          anchoredTargets.push(anchor.target)
+        else if anchor.highlights?
+          deadHighlights.push(anchor.highlights)
+          delete anchor.highlights
+          delete anchor.range
+      else
+        self.anchors.push(anchor)
+
+    deadHighlights = Array::concat(deadHighlights...)
+    raf -> highlighter.removeHighlights(deadHighlights)
+
+    for target in annotation.target when target not in anchoredTargets
+      anchor = locate(target).then(highlight)
+      anchors.push(anchor)
+
+    Promise.all(anchors).then(sync)
+
+    return annotation
+
+  createAnnotation: (annotation = {}) ->
+    self = this
+    root = @element[0]
+
+    ranges = @selectedRanges ? []
+    @selectedRanges = null
+
+    getSelectors = (range) ->
+      options = {
+        cache: self.anchoringCache
+        ignoreSelector: '[class^="annotator-"]'
+      }
+      return self.anchoring.describe(root, range, options)
+
+    setDocumentInfo = ({metadata, uri}) ->
+      annotation.uri = uri
+      if metadata?
+        annotation.document = metadata
+
+    setTargets = ([info, selectors]) ->
+      source = info.uri
+      annotation.target = ({source, selector} for selector in selectors)
+
+    info = this.getDocumentInfo().then(setDocumentInfo)
+    selectors = Promise.all(ranges.map(getSelectors))
+    targets = Promise.all([info, selectors]).then(setTargets)
+
+    targets.then(=> this.setupAnnotation(annotation))
+    targets.then(=> this.publish('beforeAnnotationCreated', [annotation]))
 
-  createAnnotation: ->
-    annotation = super
-    this.plugins.CrossFrame.sync([annotation])
     annotation
 
   createHighlight: ->
-    annotation = $highlight: true
-    this.publish 'beforeAnnotationCreated', [annotation]
-    this.plugins.CrossFrame.sync([annotation])
-    annotation
+    return this.createAnnotation({$highlight: true})
+
+  deleteAnnotation: (annotation) ->
+    anchors = []
+    targets = []
+    unhighlight = []
+
+    for anchor in @anchors
```

Comment:
I think `Array.prototype.filter` is supported for IE9 and up, so it might be clearer to use that here.